### PR TITLE
feat(coachmark): add coachmark popover

### PR DESCRIPTION
**Pull request by nickschaap**
## Description

This PR introduces the Coachmark Popover pattern into SWC.

TODO: Update README

## Related issue(s)

#2613

-

## Motivation and context

Coachmark popovers provide consumers with a mechanism to add learn content to their applications. They are often used to provide timely notifications/reminders or as building blocks for full product tours.

## How has this been tested?

-   [x] Run newly created unit tests
-   [x] Manual testing through the SWC storybook

## Screenshots (if appropriate)

image

## Types of changes

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)
-   [ ] Chore (minor updates related to the tooling or maintenance of the repository, does not impact compiled assets)

## Checklist

-   [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
-   [x] My code follows the code style of this project.
-   [ ] If my change required a change to the documentation, I have updated the documentation in this pull request.
-   [x] I have read the **[CONTRIBUTING](<(https://github.com/adobe/spectrum-web-components/blob/main/CONTRIBUTING.md)>)** document.
-   [x] I have added tests to cover my changes.
-   [ ] All new and existing tests passed.
-   [ ] I have reviewed at the Accessibility Practices for this feature, see: [Aria Practices](https://www.w3.org/TR/wai-aria-practices/)

## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.

### CODE REVIEW

File: packages/coachmark/src/spectrum-coachmark-popover-content.css [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+/* THIS FILE IS MACHINE GENERATED. DO NOT EDIT */
+@keyframes pulse {
```
Comment (nickschaap) [inline packages/coachmark/src/spectrum-coachmark-popover-content.css:14]: Not sure how to adjust the `spectrum-config` to exclude these animations here

File: packages/coachmark/src/CoachmarkPopover.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            this.addEventListener(
+                'sp-opened',
+                () => {
+                    res();
+                },
+                { once: true }
+            );
+            this.closeCoachmarkFunction = await openOverlay(
+                this.target,
+                'custom',
```
Comment (nickschaap) [inline packages/coachmark/src/CoachmarkPopover.ts:140]: Is the correct trigger interaction here?

File: packages/coachmark/src/CoachmarkPopover.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    override render(): TemplateResult {
+        return html`
+            <div id="target">
+                <slot></slot>
+                ${when(
+                    this.open,
+                    () => html`
+                        <div id="coachmark-wrapper">
+                            <slot name="pulse">
+                                <sp-coachmark></sp-coachmark>
```
Comment (nickschaap) [inline packages/coachmark/src/CoachmarkPopover.ts:184]: Should we be providing an `sp-coachmark` by default for convenience or should we leave this up to the consumer?

File: packages/coachmark/test/coachmark-popover.test.ts [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                primary-cta="Got it!"
+            ></sp-coachmark-popover-content>
+        </sp-coachmark-popover>
+    `;
+}
+
+describe('Coachmark Popover', () => {
+    testForLitDevWarnings(
+        async () => await fixture<CoachmarkPopover>(template(true))
+    );
```
Comment (nickschaap) [inline packages/coachmark/test/coachmark-popover.test.ts:55]: For some reason this test is giving me the following error:
```
TypeError: Attempted to wrap warn which is already wrapped
```

**Comment by nickschaap** [comment]
@Westbrook Hey just checking in. I believe this is ready for review. I had some questions on what may be causing the build to fail on CI that I was hoping you could help me out with

**Comment by Westbrook** [comment]
You’ve added stories, that adds VRTs, and the new tests need to be added. I’ll take a pass at the code tomorrow or so.

**Comment by Westbrook** [comment]
I’ll also look at the perf check pass. I thought we fixed that issue, but maybe it’s a new one.

**Comment by nickschaap** [comment]
@Westbrook I rebase to main and updated the baseline hash. I'm sort of stuck on the failing unit test. That seems to be complaining about wrapping `console.warn` through `testForLitDevWarnings` helper more than once. This helper restores the stub after it uses it so I'm not sure what could be going on here.

File: packages/coachmark/test/coachmark.test.ts [hunk 5]
```diff
@@ -14,7 +14,7 @@ import { elementUpdated, expect, fixture, html } from '@open-wc/testing';
 
 import '@spectrum-web-components/coachmark/sp-coachmark.js';
 import { Coachmark } from '@spectrum-web-components/coachmark';
-import { testForLitDevWarnings } from '../../../test/testing-helpers.js';
+import { testForLitDevWarnings } from '../../../test/testing-helpers';
```
Comment (Westbrook) [inline packages/coachmark/test/coachmark.test.ts:17]: Can you try with the `.js` added back?

Comment (nickschaap) [reply packages/coachmark/test/coachmark.test.ts]: I adjusted it locally but I'm still running into a test failure in `coachmark-popover.test.ts`

File: packages/coachmark/test/coachmark-popover.test.ts (hunk 4 above)
Comment (Westbrook) [reply packages/coachmark/test/coachmark-popover.test.ts:55]: I can't figure out why this is doing this, or why I can't cause other things to do this, or how I specifically prevented this when I saw it in the past, _BUT_ it seems that this patch makes it go away...

[controllerWrap.patch](https://github.com/adobe/spectrum-web-components/files/11245228/controllerWrap.patch)

Can you try this locally to confirm? I'd commit it separately, but I'm not a fan of commits that I can't test, and as I can't cause this error in other packages.

**Comment by Westbrook** [comment]
Spectrum CSS is doing some funny business with this and [Core Tokens](https://github.com/adobe/spectrum-css/pull/1992).  Specifically splitting this into [Indicator](https://pr-1992--spectrum-css.netlify.app/coachindicator.html) and [Mark](https://pr-1992--spectrum-css.netlify.app/coachmark.html). It might not change much _our_ delivery, but there are some sizable design changes as this move from "contribution" to "official".

So, as much as I'm sorry this has gotten help up in various ways, it looks like it might be for the betterment of all! 🤞🏼

**Comment by Westbrook** [comment]
Closing in favor of #3639

**Pull request closed without merging**